bpf/lib/common.h

```diff
@@ -245,9 +249,20 @@ struct endpoint_info {
 	__u32		pad[4];
 };
 
-struct egress_info {
-	__u32 egress_ip;
-	__u32 tunnel_endpoint;
+struct egress_policy_key {
+	struct bpf_lpm_trie_key lpm_key;
+	__be32 saddr;
+	__be32 daddr;
+};
```

Comment:
Shouldn't be an issue as I renamed the old `cilium_egress_v4` to `cilium_egress_policy_v4` (in order to have consistent naming with the other related map, `cilium_egress_ct_v4`)

Reply:
Renaming also typically requires special consideration for upgrades/downgrades. We currently don't wait for the Cilium Egress Gateway Policies to be synced from k8s to restore and regenerate the endpoints. So on both upgrades and downgrades we will at least have a short connectivity blip.

Did you try to perform an upgrade+downgrade with entries in the map?

Reply:
> We currently don't wait for the Cilium Egress Gateway Policies to be synced from k8s to restore and regenerate the endpoints.

I haven't considered that. In this case I think too that with the current state of things there will be connection breakages with upgrades and downgrades (although I haven't tested yet this with entries in the map).

For downgrades I'm not sure there's much we can do given that it's not always possible to map the state of an HA configuration to a non HA one (i.e. it's not clear what we should do when we have multiple gateways for a given policy and downgrade to 1.10 which only supports a single gateway). Moreover egress gateway is still marked as beta, so that may be acceptable?

On the other hand, for upgrades it should be possible to migrate the old map to the new one while the agent has not synced with k8s (but I haven't properly thought about that yet).

Reply:
> For downgrades I'm not sure there's much we can do given that it's not always possible to map the state of an HA configuration to a non HA one (i.e. it's not clear what we should do when we have multiple gateways for a given policy and downgrade to 1.10 which only supports a single gateway).

Downgrades are most likely to happen before users start using the new egress gateway feature. That is, I'd expect downgrades to happen shortly after upgrades, in case the upgrade has issues. Given that, the egress gateway configuration at the time of the downgrade will likely not be an HA one.

> On the other hand, for upgrades it should be possible to migrate the old map to the new one while the agent has not synced with k8s (but I haven't properly thought about that yet).

Probably safe to just instruct the agent to wait for the k8s sync?

Reply:
> Probably safe to just instruct the agent to wait for the k8s sync?

I don't have much context on why we are currently not waiting for k8s to sync but if there's nothing in particular preventing us from doing so then yes, we can change the code to wait for it :+1:

Reply:
I'd expect it to just be an oversight. It will slow the agent startup a bit, but probably still the right thing to do. Probably best done as a separate PR though, to be able to backport it to v1.10 (that way the downgrade should be fixed as well).